src/coverlet.core/Symbols/CecilSymbolHelper.cs

```diff
@@ -0,0 +1,290 @@
+//
+// This class is based heavily on the work of the OpenCover
+// team in OpenCover.Framework.Symbols.CecilSymbolManager
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+using Coverlet.Core.Extensions;
+
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Collections.Generic;
+
+namespace Coverlet.Core.Symbols
+{
+    public static class CecilSymbolHelper
+    {
+        private const int StepOverLineCode = 0xFEEFEE;
+        private static readonly Regex IsMovenext = new Regex(@"\<[^\s>]+\>\w__\w(\w)?::MoveNext\(\)$", RegexOptions.Compiled | RegexOptions.ExplicitCapture);
+        public static List<BranchPoint> GetBranchPoints(MethodDefinition methodDefinition)
+        {
+            var list = new List<BranchPoint>();
+            GetBranchPoints(methodDefinition, list);
+            return list;
+        }
+        private static void GetBranchPoints(MethodDefinition methodDefinition, List<BranchPoint> list)
+        {
+            if (methodDefinition == null) 
+                return;
+            try
+            {
+                UInt32 ordinal = 0;
+                var instructions = methodDefinition.Body.Instructions;
+                
+                // if method is a generated MoveNext skip first branch (could be a switch or a branch)
+                var skipFirstBranch = IsMovenext.IsMatch(methodDefinition.FullName);
+
+                foreach (var instruction in instructions.Where(instruction => instruction.OpCode.FlowControl == FlowControl.Cond_Branch))
+                {
+                    if (skipFirstBranch)
+                    {
+                        skipFirstBranch = false;
+                        continue;
+                    }
+
+                    if (BranchIsInGeneratedFinallyBlock(instruction, methodDefinition)) 
+                        continue;
+
+                    var pathCounter = 0;
+
+                    // store branch origin offset
+                    var branchOffset = instruction.Offset;
+                    var closestSeqPt = FindClosestInstructionWithSequencePoint(methodDefinition.Body, instruction).Maybe(i => methodDefinition.DebugInformation.GetSequencePoint(i));
+                    var branchingInstructionLine = closestSeqPt.Maybe(x => x.StartLine, -1);
+                    var document = closestSeqPt.Maybe(x => x.Document.Url);
+
+                    if (null == instruction.Next)
+                        return;
+
+                    if (!BuildPointsForConditionalBranch(list, instruction, branchingInstructionLine, document, branchOffset, pathCounter, instructions, ref ordinal, methodDefinition)) 
+                        return;
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException(
```

Comment:
We should have a sane fallback instead of throwing an exception.

Reply:
This is the only change I haven't yet made from this review.  Do you have any suggestions?

Reply:
Perhaps put the `try...catch` inside the loop and simply `continue` to ignore the error. I don't see much point throwing an exception because the developer most likely won't know what/how to fix it. IMO better to have some missing coverage info than to have the entire thing fail with no directions on how to fix it. Also if this file was pulled from OpenCover I think it'll be right to attribute it to the project